File: ruby/lib/epostak/resources/payloads.rb
# frozen_string_literal: true

require "stringio"

module EPostak
  module Resources
    # Payload Assistant helpers for creating send payloads.
    class Payloads
      def initialize(http)
        @http = http
      end

      def extract(file_path_or_io, mime_type, file_name: "document", corrected_fields: nil)
        @http.request_multipart("/payloads/extract", [
          { field: "file", io: resolve_io(file_path_or_io), mime_type: mime_type, filename: file_name }
        ], fields: multipart_corrections(corrected_fields))
      end

      def extract_batch(files)
        parts = files.map do |f|
          {
            field: "files",
            io: resolve_io(f[:file]),
            mime_type: f[:mime_type],
            filename: f[:file_name] || "document"
          }
        end
        @http.request_multipart("/payloads/extract/batch", parts)
      end

      def parse(xml)
        @http.request(:post, "/payloads/parse", body: { xml: xml })
      end

      def convert(input_format:, output_format:, document:)
        @http.request(
          :post,
          "/payloads/convert",
          body: { input_format: input_format, output_format: output_format, document: document }
        )
      end

      def validate(body)
        @http.request(:post, "/payloads/validate", body: body)
      end

      private

      def multipart_corrections(corrected_fields)
        corrected_fields.nil? ? {} : { "fields" => JSON.generate(corrected_fields) }
      end

      def resolve_io(file_path_or_io)
        file_path_or_io.is_a?(String) ? File.open(file_path_or_io, "rb") : file_path_or_io
      end
    end
  end
end

File: python/src/epostak/resources/payloads.py
"""Payload Assistant resource."""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from epostak.resources.documents import _BaseResource


class PayloadsResource(_BaseResource):
    """Create, parse, convert, and validate send payloads."""

    def extract(
        self,
        file: bytes,
        mime_type: str,
        file_name: str = "document",
    ) -> Dict[str, Any]:
        files = [("file", (file_name, file, mime_type))]
        return self._request("POST", "/payloads/extract", files=files)

    def extract_batch(self, files: List[Dict[str, Any]]) -> Dict[str, Any]:
        upload_files = []
        for item in files:
            upload_files.append(
                ("files", (item.get("file_name", "document"), item["file"], item["mime_type"]))
            )
        return self._request("POST", "/payloads/extract/batch", files=upload_files)

    def parse(self, xml: str) -> Dict[str, Any]:
        return self._request("POST", "/payloads/parse", json={"xml": xml})

    def convert(self, input_format: str, output_format: str, document: Any) -> Dict[str, Any]:
        return self._request(
            "POST",
            "/payloads/convert",
            json={
                "input_format": input_format,
                "output_format": output_format,
                "document": document,
            },
        )

    def validate(self, body: Dict[str, Any]) -> Dict[str, Any]:
        return self._request("POST", "/payloads/validate", json=body)
